src/App.tsx

```diff
@@ -1,14 +1,78 @@
 /* eslint-disable max-len */
-import React from 'react';
+import React, { useEffect, useState, useMemo } from 'react';
 import 'bulma/css/bulma.css';
 import '@fortawesome/fontawesome-free/css/all.css';
 
 import { TodoList } from './components/TodoList';
 import { TodoFilter } from './components/TodoFilter';
 import { TodoModal } from './components/TodoModal';
 import { Loader } from './components/Loader';
+import { getTodos } from './api';
+import { Todo } from './types/Todo';
+import { Filter } from './types/Filter';
 
 export const App: React.FC = () => {
+  const [filter, setFilter] = useState<Filter>(Filter.All);
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [activeTodo, setActiveTodo] = useState<Todo | null>(null);
+  const [query, setQuery] = useState('');
+  const [isLoadingTodos, setIsLoadingTodos] = useState(true);
+
+  const cachedTodos = useMemo(() => getTodos(), []);
+
+  useEffect(() => {
+    setIsLoadingTodos(true);
+
+    switch (filter) {
+      case Filter.All:
+        cachedTodos
+          .then(allTodos =>
+            allTodos.filter(todo =>
+              todo.title.toLowerCase().includes(query.toLowerCase()),
+            ),
+          )
+          .then(setTodos)
+          .finally(() => setIsLoadingTodos(false));
+        break;
+      case Filter.Active:
+        cachedTodos
+          .then(allTodos =>
+            allTodos.filter(
+              todo =>
+                !todo.completed &&
+                todo.title.toLowerCase().includes(query.toLowerCase()),
+            ),
+          )
+          .then(setTodos)
+          .finally(() => setIsLoadingTodos(false));
+        break;
+      case Filter.Completed:
+        cachedTodos
+          .then(allTodos =>
+            allTodos.filter(
+              todo =>
+                todo.completed &&
+                todo.title.toLowerCase().includes(query.toLowerCase()),
+            ),
+          )
```

Comment:
it looks strange and complex, just move it to some function that will return the filtered array
`const visibleTodos = getVisibleTodos(todos, filter, query)`

and then use that array everywhere